dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,497 @@
+      preconditionOrValues
+    );
+  }
+
+  /**
+   * Returns a Promise that resolves when there are no more pending writes. It
+   * never fails. Calling this method will send all requests.
+   *
+   * The promise resolves immediately if there are no pending writes. Otherwise,
+   * the Promise waits for all previously issued writes, but it does not wait
+   * for writes that were added after the method is called. If you want to wait
+   * for additional writes, call `waitForPendingWrites()` again.
+   *
+   * @return {Promise<void>} A promise that resolves when there are no more
+   * pending writes.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter.create(documentRef, {foo: 'bar'});
+   * bulkWriter.update(documentRef2, {foo: 'bar'});
+   * bulkWriter.delete(documentRef3);
+   * await waitForPendingWrites().then(() => {
+   *   console.log('Successfully executed all writes');
+   * });
+   */
+  async waitForPendingWrites(): Promise<void> {
+    const trackedBatches = this.batchQueue;
+    const writePromises = trackedBatches.map(batch => batch.awaitBatch());
+    this.sendBatches();
+    await Promise.all(writePromises);
+  }
+
+  /**
+   * Return the first eligible batch that can hold a write to the provided
+   * reference, or creates one if no eligible batches are found.
+   */
+  private getEligibleBatch(ref: DocumentReference): BulkCommitBatch {
+    let batch = null;
+    let logWarning = this.batchQueue.length > 0;
+    for (const bulkBatch of this.batchQueue) {
+      if (!bulkBatch.refsInBatch.has(ref)) {
+        logWarning = false;
+      }
+      if (bulkBatch.canAddDoc(ref)) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        console.warn(
+          '[BulkWriter]',
+          'Writing to the same document multiple times will slow down BulkWriter. ' +
+            'Write to unique documents in order to maximize throughput.'
+        );
+      }
+      return this.createNewBatch();
+    } else {
+      return batch;
+    }
+  }
+
+  private createNewBatch(): BulkCommitBatch {
+    const newBatch = new BulkCommitBatch(this.getId(), this.firestore.batch());
+
+    this.batchQueue.push(newBatch);
+    return newBatch;
+  }
+
+  private getId(): number {
+    const id = this.currentId;
+    if (this.currentId === Number.MAX_SAFE_INTEGER) {
+      this.currentId = 0;
+    } else {
+      this.currentId++;
+    }
+    return id;
+  }
+
+  /**
+   * Attempts to send batches starting from the front of the BatchQueue until a
+   * batch cannot be sent.
+   *
+   * After a batch is complete, try sending batches again.
+   */
+  private sendBatches(): void {
+    let counter = 0;
+    for (const batch of this.batchQueue) {
+      // Only send up until the first batch that cannot be sent.
+      if (!this.isBatchSendable(batch)) {
+        break;
+      }
+
+      this.refsInFlight = new Set([...this.refsInFlight, ...batch.refsInBatch]);
```

Comment:
Declaring a new `Set` and using the spread operator is obscuring the intent of what you're trying to communicate and could be done more efficiently. What you're ultimately trying to do is append `refsInFlight` with the contents in the batch. With `Set`s this is accomplished through the `add()` method. You're also allocating new memory with no clear benefit.

I'd suggest updating this to something more like `batch.refsInBatch.forEach(this.refsInFlight.add);`.

Reply:
done.